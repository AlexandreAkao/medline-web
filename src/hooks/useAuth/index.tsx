import { useContext } from 'react';
import { isEmpty } from 'ramda';

import AuthContext from 'contexts/AuthContext';

export const useAuth = (): IAuthContext => {
  const context = useContext(AuthContext);

  if (isEmpty(context)) {
    throw new Error('Your component must be wrapped in a AuthProvider');
  }

  return context;
};
