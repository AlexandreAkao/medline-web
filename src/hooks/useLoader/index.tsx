import { useContext } from 'react';
import { isEmpty } from 'ramda';

import LoaderContext from 'contexts/LoaderContext';

export const useLoader = (): ILoaderContext => {
  const context = useContext(LoaderContext);

  if (isEmpty(context)) {
    throw new Error('Your component must be wrapped in a LoaderProvider');
  }

  return context;
};
