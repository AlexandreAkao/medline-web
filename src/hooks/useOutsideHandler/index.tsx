import { MutableRefObject, useEffect } from 'react';

function useOutsideHandler<T extends HTMLElement>(
  ref: MutableRefObject<T | null>,
  action?: (event: MouseEvent) => void,
) {
  useEffect(() => {
    function handleClickOutside(event: MouseEvent) {
      if (ref.current && !ref.current.contains(event.target as Node)) {
        action && action(event);
      }
    }

    document.addEventListener('mousedown', handleClickOutside);
    return () => {
      document.removeEventListener('mousedown', handleClickOutside);
    };
  }, [action, ref]);
}

export default useOutsideHandler;
