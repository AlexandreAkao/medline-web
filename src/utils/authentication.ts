import { isEmpty } from 'ramda';

export const getUserInfo = () => {
  const user = JSON.parse(localStorage.getItem('user') ?? '{}');

  return isEmpty(user) ? null : (user as IUser);
};

export const geToken = () => localStorage.getItem('token');
