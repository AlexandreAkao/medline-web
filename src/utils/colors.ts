export const getColorAlpha = (color: string, alpha = 1) => {
  const alphaLevel = Math.round(255 * alpha).toString(16);
  return color + alphaLevel;
};
