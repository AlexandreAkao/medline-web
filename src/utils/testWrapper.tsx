import { render } from '@testing-library/react';
import React from 'react';
import { Router } from 'react-router-dom';
import { createMemoryHistory, MemoryHistory } from 'history';

import { AuthProvider } from 'providers/AuthProvider';
import { LoaderProvider } from 'providers/LoaderProvider';

export function RouterProviders({
  children,
  route = '/',
  history = createMemoryHistory({ initialEntries: [route] }),
}: IChildren & { route?: string; history?: MemoryHistory }) {
  return (
    <Router location={history.location} navigator={history}>
      {children}
    </Router>
  );
}

export const renderWithRouter = (
  component: React.ReactNode,
  { route = '/', history = createMemoryHistory({ initialEntries: [route] }) } = {},
) => {
  return {
    ...render(
      <Router location={history.location} navigator={history}>
        {component}
      </Router>,
    ),
    history,
  };
};

export const renderWithProvider = (
  component: React.ReactNode,
  { route = ['/'], history = createMemoryHistory({ initialEntries: route }) } = {},
) => {
  return {
    ...render(
      <Router location={history.location} navigator={history}>
        <LoaderProvider>
          <AuthProvider>{component}</AuthProvider>,
        </LoaderProvider>
      </Router>,
    ),
    history,
  };
};
