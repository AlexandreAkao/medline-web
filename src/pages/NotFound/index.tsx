import { useNavigate } from 'react-router-dom';

import Button from 'components/Button';
import { useAuth } from 'hooks/useAuth';
import {
  GlobalStyle,
  NotFoundContainer,
  NotFoundMessage,
  NotFoundSubTitle,
  NotFoundTitle,
} from 'pages/NotFound/styles';

function NotFound() {
  const navigate = useNavigate();
  const { isEmployee } = useAuth();

  const handleGoToHome = () => navigate(isEmployee ? '/employee/patiente-care' : '/');

  return (
    <NotFoundContainer>
      <GlobalStyle />
      <NotFoundTitle>Oops!</NotFoundTitle>
      <NotFoundSubTitle>404 - Página não encontrada</NotFoundSubTitle>
      <NotFoundMessage role="article">
        A página que você está procurando pode ter sido removida teve seu nome alterado ou está temporariamente
        indisponível.
      </NotFoundMessage>
      <Button onClick={handleGoToHome}>Voltar para Home</Button>
    </NotFoundContainer>
  );
}

export default NotFound;
