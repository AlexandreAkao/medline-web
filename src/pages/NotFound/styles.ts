import styled, { createGlobalStyle } from 'styled-components';

import colors from 'styles/colors';
import { getColorAlpha } from 'utils/colors';

export const GlobalStyle = createGlobalStyle`
  body {
    background-color: ${colors.primary.lighter};
  }
`;

export const NotFoundContainer = styled.div`
  height: 100%;
  overflow: auto;
  display: flex;
  justify-content: center;
  align-items: center;
  flex-direction: column;
`;

export const NotFoundTitle = styled.h1`
  font-size: 15rem;
  font-weight: 800;
  margin-bottom: 2rem;
  background-image: linear-gradient(90deg, ${colors.primary.light}, ${colors.primary.darker});
  -webkit-background-clip: text;
  -webkit-text-fill-color: transparent;
  position: relative;

  ::after {
    background: none;
    display: inline-block;
    content: 'Oops!';
    left: 0;
    position: absolute;
    text-shadow: -4px 4px 3px ${getColorAlpha(colors.black.soft, 0.5)};
    top: 0;
    z-index: -1;
  }
`;

export const NotFoundSubTitle = styled.span`
  font-size: 1.5rem;
  font-weight: 700;
  margin-bottom: 1rem;
`;

export const NotFoundMessage = styled.span`
  font-size: 1.25rem;
  width: 35%;
  text-align: center;
  margin-bottom: 1.5rem;
`;
