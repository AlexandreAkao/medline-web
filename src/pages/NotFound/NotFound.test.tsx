import { screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { vi } from 'vitest';

import { renderWithProvider } from 'utils/testWrapper';
import NotFound from 'pages/NotFound';
import * as useAuthHook from 'hooks/useAuth';

describe('NotFound', () => {
  const renderComponent = () => {
    return renderWithProvider(<NotFound />, { route: ['/any-wrong-url'] });
  };

  it('should match with snapshot', () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  it('should render correctly', () => {
    renderComponent();

    const titleElement = screen.getByText('Oops!');
    const subTitleElement = screen.getByText('404 - Página não encontrada');
    const messageElement = screen.getByRole('article');
    const buttonElement = screen.getByRole('button');

    expect(titleElement).toBeInTheDocument();
    expect(subTitleElement).toBeInTheDocument();
    expect(messageElement).toBeInTheDocument();
    expect(buttonElement).toBeInTheDocument();
  });

  describe('should redirect to home page', () => {
    it('when is patiente', async () => {
      vi.spyOn(useAuthHook, 'useAuth').mockImplementation(
        () =>
          ({
            isEmployee: false,
          } as IAuthContext),
      );

      const { history } = renderComponent();

      expect(history.location.pathname).toBe('/any-wrong-url');

      const buttonElement = screen.getByRole('button');
      await userEvent.click(buttonElement);

      expect(buttonElement).toBeInTheDocument();

      expect(history.location.pathname).toBe('/');
    });

    it('when is employee', async () => {
      vi.spyOn(useAuthHook, 'useAuth').mockImplementation(
        () =>
          ({
            isEmployee: true,
          } as IAuthContext),
      );

      const { history } = renderComponent();

      expect(history.location.pathname).toBe('/any-wrong-url');

      const buttonElement = screen.getByRole('button');
      await userEvent.click(buttonElement);

      expect(buttonElement).toBeInTheDocument();

      expect(history.location.pathname).toBe('/employee/patiente-care');
    });
  });
});
