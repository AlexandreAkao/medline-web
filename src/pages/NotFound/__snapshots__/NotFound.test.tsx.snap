// Vitest Snapshot v1

exports[`NotFound > should match with snapshot 1`] = `
<div>
  <div
    class="sc-gsnTZi dGuJDO"
  >
    <h1
      class="sc-dkzDqf eUVvWH"
    >
      Oops!
    </h1>
    <span
      class="sc-hKMtZM jeCfWs"
    >
      404 - Página não encontrada
    </span>
    <span
      class="sc-eCYdqJ gMWbyh"
      role="article"
    >
      A página que você está procurando pode ter sido removida teve seu nome alterado ou está temporariamente indisponível.
    </span>
    <button
      class="sc-bczRLJ bJBahq"
      type="button"
    >
      Voltar para Home
    </button>
  </div>
  ,
</div>
`;
